File: core/ast/ast.ts
export type Node = Program | Statement | Expression;

// The root of the entire tree
export interface Program {
  type: "Program";
  body: Statement[];
}

// Actions like "let x = 5"
export interface VariableDeclaration {
  type: "VariableDeclaration";
  identifier: string;
  value: Expression;
}

// Math or values like "5 + 10"
export interface BinaryExpression {
  type: "BinaryExpression";
  left: Expression;
  operator: string;
  right: Expression;
}

export interface Literal {
  type: "Literal";
  value: number | string;
}

export interface Comment {
  type: "Comment";
  value: string;
}

export interface String {
  type: "String";
  value: string;
}

export interface Double {
  type: "Double";
  value: number;
}

export interface Integer {
  type: "Integer";
  value: number;
}

export interface Function {
  type: "Function";
  name: string;
  parameters: string[];
  body: Statement[];
}

export interface BooleanLiteral {
  type: "BooleanLiteral";
  value: boolean;
}

export interface Identifier {
  type: "Identifier";
  name: string;
}

export interface PrintStatement {
  type: "PrintStatement";
  expression: Expression; // Changed from 'identifier' to 'expression'
}

export interface ArrayLiteral {
  type: "ArrayLiteral";
  elements: Expression[];
}

export interface CallExpression {
  type: "CallExpression";
  callee: string;
  arguments: Expression[];
}

export interface Dot {
  type: "Dot";
  name: string
}

export interface ComparisonExpression {
  type: "ComparisonExpression";
  left: Expression;
  operator: string; // e.g., "==", "!=", "<", ">", "<=", ">="
  right: Expression;
}

export interface ConditionalExpression {
  type: "ConditionalExpression";
  test: Expression; // The condition to evaluate
  consequent: Statement[]; // Statements to execute if condition is true
  alternate: Statement[]; // Statements to execute if condition is false (optional)
}

export interface LoopExpression {
  type: "LoopExpression";
  test: Expression; // The condition to evaluate before each iteration
  body: Statement[]; // Statements to execute in each iteration
}

export interface IndexExpression {
  type: "IndexExpression";
  object: Expression; // The array or object being indexed
  index: Expression; // The index or key being accessed
}

export type Statement =
  | VariableDeclaration
  | PrintStatement
  | Comment
  | Function
  | Expression
  | String
  | BooleanLiteral
  | CallExpression
  | ComparisonExpression
  | ConditionalExpression
  | LoopExpression
  | ArrayLiteral
  | IndexExpression
  | Double
  | Integer
  | Dot;
export type Expression =
  | BinaryExpression
  | Literal
  | Identifier
  | CallExpression
  | String
  | BooleanLiteral
  | ComparisonExpression
  | ConditionalExpression
  | LoopExpression
  | ArrayLiteral
  | IndexExpression
  | Double
  | Integer
  | Dot;

File: src/core/ast/ast.go
package ast

// Node is the base interface every AST node implements
type Node interface{}

// Program is the top-level AST node
type Program struct {
	Body []Node
}

// VariableDeclaration: let x: int = 5; or const x: int = 5;
type VariableDeclaration struct {
	Identifier   string
	DeclaredType string
	Value        Node
	Mutable      bool // true for let, false for const
}

// Function: func add(a: int, b: int): int { ... }
type Function struct {
	Name       string
	Params     []Param
	ReturnType string
	Body       []Node
}

type Param struct {
	Name string
	Type string
}

// ReturnStatement: return x + 1;
type ReturnStatement struct {
	Value Node
}

// PrintStatement: print "hello";
type PrintStatement struct {
	Expression Node
}

// IfStatement: if x > 5 { ... } else { ... }
type IfStatement struct {
	Condition  Node
	Consequent []Node
	Alternate  []Node
}

// WhileLoop: while x < 10 { ... }
type WhileLoop struct {
	Condition Node
	Body      []Node
}

// Comment: # this is a comment
type Comment struct {
	Value string
}

// BinaryExpression: x + y, x == y, etc.
type BinaryExpression struct {
	Left     Node
	Operator string
	Right    Node
}

// CallExpression: add(3, 4)
type CallExpression struct {
	Callee    string
	Arguments []Node
}

// Increment: x++
type Increment struct {
	Identifier string
}

// Decrement: x--
type Decrement struct {
	Identifier string
}

// Identifier: a variable reference
type Identifier struct {
	Name string
}

// LiteralInt: 42
type LiteralInt struct {
	Value int64
}

// LiteralFloat: 3.14
type LiteralFloat struct {
	Value float64
}

// LiteralString: "hello"
type LiteralString struct {
	Value string
}

// LiteralBool: true / false
type LiteralBool struct {
	Value bool
}

// IndexExpression: arr[0]
type IndexExpression struct {
	Object Node
	Index  Node
}

// Assignment: x = 5
type Assignment struct {
	Identifier string
	Value      Node
}

// ExpressionStatement: wrapper for expressions used as statements
type ExpressionStatement struct {
	Expression Node
}

// UnaryExpression: -5, !true
type UnaryExpression struct {
	Operator string
	Operand  Node
}
